### PR TITLE
[WB-7574] Improve xgboost wandb_callback (#2929)

### PR DESCRIPTION
**Pull request by raubitsj**
From: https://github.com/wandb/client/pull/2929

Description
-----------
What does the PR do?

This PR improves and extends the `wandb_callback` for XGBoost. Here's the list of features that the new `WandbCallback` has:
    
- log the booster model configuration to Weights & Biases
- log evaluation metrics collected by XGBoost, such as rmse, accuracy, etc to Weights & Biases
- log training metric collected by XGBoost (if you provide training data to eval_set)
- log the best score and the best iteration
- save and upload your trained model to Weights & Biases Artifacts (when `log_model = True`)
- log feature importance plot when `log_feature_importance=True` (default).
- Capture the best eval metric in `wandb.summary` when `define_metric=True` (default).

Here's an example dashboard: https://wandb.ai/ayut/credit_scorecard

Testing
-------
How was this PR tested?

`yea` test.

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/integration/xgboost/xgboost.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        self.log_model: bool = log_model
+        self.log_feature_importance: bool = log_feature_importance
+        self.importance_type: str = importance_type
+        self.define_metric: bool = define_metric
+
+        with wb_telemetry.context() as tel:
+            tel.feature.xgboost_wandb_callback = True
+
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before WandbCallback()")
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:112]:
```suggestion
        if wandb.run is None:
            raise wandb.Error("You must call wandb.init() before WandbCallback()")

        with wb_telemetry.context() as tel:
            tel.feature.xgboost_wandb_callback = True
```

cant add telemetry without the run. want to reorder these.

File: functional_tests/xgboost/xgboost_reg.yea [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  - :op:>:
+    - :wandb:runs[0][summary][validation_0-rmse][min]
+    - 0.0
+  - :op:>:
+    - :wandb:runs[0][summary][validation_1-rmse][min]
+    - 0.0
+  - :wandb:runs[0][exitcode]: 0
+  - :op:contains:
+    - :wandb:runs[0][telemetry][3]  # feature
+    - 31  # xgboost_wandb_callback
```
Comment (raubitsj) [inline functional_tests/xgboost/xgboost_reg.yea:29]:
```suggestion
    - 31  # xgboost_wandb_callback
```
